cumulus/parachains/runtimes/coretime/coretime-rococo/src/lib.rs

```diff
@@ -740,8 +740,15 @@ impl_runtime_apis! {
 				}
 
 				fn reserve_transferable_asset_and_dest() -> Option<(Asset, Location)> {
-					// Reserve transfers are disabled
-					None
+					// Coretime chain can reserve transfer regions to some random parachain.
+					let raw_region_id = 42;
+					Some((
+						Asset {
+							fun: NonFungible(Index(raw_region_id)),
+							id: AssetId(Location::new(0, [PalletInstance(50)]))
```

Comment:
This will still fail since we don't support minting regions, resulting in setting the weight to the max value.

I added 'region minting' here: https://github.com/paritytech/polkadot-sdk/pull/3553, however, that would still disallow this.

Reply:
@franciscoaguirre Any ideas on what we might do in this regard?

Reply:
cc @seadanda

Reply:
@Szegoo is this still an issue? I think you need to create a region with `raw_region_id=42` first, so maybe some broker pallet magic, e.g.:
```
        let caller: T::AccountId = whitelisted_caller();
		T::Currency::set_balance(
			&caller.clone(),
			T::Currency::minimum_balance().saturating_add(20u32.into()),
		);

        /// TODO: maybe some more magic e.g. `setup_and_start_sale` in the pallet_broker's benchmarks
		let region = Broker::<T>::do_purchase(caller.clone(), 10u32.into())?;
```

or alternatively, simply inserting some data could do the trick:
```
let raw_region_id = 42;
let region = RegionRecord<AccountId, Balance> {
	end: ...,
	owner: ...,
	paid: ...,
};
Regions::<T>::insert(&raw_region_id, &region);
```

Reply:
Inside `pallet-xcm` when running the benchmark for reserve transfers it will attempt to deposit the region to an account before transferring it.

This requires the `mint_into` to be implemented for regions.
 
A possible solution could be to do something like this: 
```rust
	#[cfg(features = "runtime-benchmarks")]
	fn mint_into(item: &Self::ItemId, who: &T::AccountId) -> DispatchResult {
		let region_id: RegionId = (*item).into();

		// a dummy value for benchmarking.
		let end = region_id.begin.saturating_add(1);
		Self::issue(region_id.core, region_id.begin, end, who.clone(), None);

		Ok(())
	}
```

However, this requires that the `runtime-benchmarks` feature propagates so that this implementation is used.

Reply:
@Szegoo I see, that `mint_into` seems kind of hacky :). I think there is simple solution without implementing `mint_into`. We already use this pattern for benchmarks elsewhere (I think I saw it elsewhere also), e.g. [`ParachainSystem::open_outbound_hrmp_channel_for_benchmarks_or_tests`](https://github.com/paritytech/polkadot-sdk/blob/master/cumulus/pallets/parachain-system/src/lib.rs#L1460-L1530). For this particular benchmarks, you just need "a created region", you don't care how was created, so I would suggest to do the same here:
- create pallet function into the `pallet_broker`'s [`utility_impls.rs`](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/broker/src/utility_impls.rs), which will simply just insert what is expected for success execution:
```
	#[cfg(feature = "runtime-benchmarks")]
	pub fn create_region_for_benchmark(region_id: RegionId, ..) {
		/// OR this way
		let region = RegionRecord<AccountId, Balance> {
			end: ...,
			owner: ...,
			paid: ...,
		};
		Regions::<T>::insert(&raw_region_id, &region);

		/// OR reuse/utilize above `fn issue() -> RegionId`
		Self::issue(....)
	}      
```  
- and then in the runtime's `fn reserve_transferable_asset_and_dest` you just add:
```
   let raw_region_id = Broker::create_region_for_benchmark(...);
```

Reply:
Yes, the problem is that `mint_into` is called from the `pallet-xcm` during benchmarking. Even if we create the region beforehand, it will still attempt to call `deposit_asset`: https://github.com/paritytech/polkadot-sdk/pull/3455/files#diff-299b81a18ba51e2c2fdaa3b6c7c7dd99ced49c9c7b8772fa1225ebecced332e0R188 (the `deposit_asset` function is calling `mint_into`). 

If not implemented it will fail and set the weight to max.

Maybe we don't have to set the weight to max if `deposit_asset` fails given that it would fail in the subsequent steps, however not sure if this is a better solution than what I proposed above.

Reply:
I'm not sure if I understood this correctly. Are you proposing that we call `create_region_for_benchmark` before executing the benchmark code? However, in that case, we still face the issue of `deposit_asset` failing during the benchmark.

Reply:
yes, `create_region_for_benchmark` ensures that region is created correctly,
and I meant that we just use `mint_into` from [here](https://github.com/paritytech/polkadot-sdk/pull/3553/files#diff-40b92c4632f430a76d4f2c6a366582cbd2e4898fb24cd1a106a20c39872ea4d1R69-R80), there is no need to split it for `#[cfg(not(features = "runtime-benchmarks"))]` and `#[cfg(features = "runtime-benchmarks")]` as you suggested [here](https://github.com/paritytech/polkadot-sdk/pull/3455#discussion_r1541564254)

Reply:
so final solution:
1. call `Broker::create_region_for_benchmark` before benchmark in the `fn reserve_transferable_asset_and_dest`, it makes sure we have a correct region
2. `AssetTransactor::deposit_asset(` in the benchmark, ensures that region belongs to `origin_location`, that triggers transfer to destination
    - yes, we need here `mint_into` from [here](https://github.com/paritytech/polkadot-sdk/pull/3553/files#diff-40b92c4632f430a76d4f2c6a366582cbd2e4898fb24cd1a106a20c39872ea4d1R69-R80) - you can copy this `mint_into` to this branch and see that it is working

I think this is clean solution. I would not change or ignore MAX for failed `AssetTransactor::deposit_asset(`.

Reply:
I think this will still fail:

1. call `Broker::create_region_for_benchmark` <- creates the region we will be reserve transferring
2. `AssetTransactor::deposit_asset` will fail since based on the code from #3553 we can't mint an already existing region: https://github.com/paritytech/polkadot-sdk/pull/3553/files#diff-40b92c4632f430a76d4f2c6a366582cbd2e4898fb24cd1a106a20c39872ea4d1R76

> I think this is clean solution. I would not change or ignore MAX for failed AssetTransactor::deposit_asset(.

Yeah, I agree, that isn't a good solution.

Maybe branching based on the feature flag might not actually be a bad solution

Reply:
What exactly will fail?

1. call `Broker::create_region_for_benchmark` will create region without owner (I see, that in [#3553](https://github.com/paritytech/polkadot-sdk/pull/3553/files#diff-a8dc0f7b2323a7a2b4fa8a038ab776271c156e96c0fc2f7d36c81be26889283dR87) is changed to Option):
```
let region = RegionRecord<AccountId, Balance> {
			end: ...,
			owner: None,
			paid: ...,
		};
Regions::<T>::insert(&raw_region_id, &region);
``` 

2. `AssetTransactor::deposit_asset` will call `mint_into` and that `ensure!(record.owner.is_none(), Error::<T>::NotAllowed);` will pass, because we set `owner: None`, so `mint_into` sets an owner with `origin_location` as AccountId, so it exactly does what we need.

3. `pallet_xcm::reserve_transfer_assets` executes `TransferAsset` XCM instruction which should end with calling `pallet_broker::do_transfer`, which just does some owner change.

so far so good, unless I am missing anything.

Maybe, I would suggest to add here: https://github.com/paritytech/polkadot-sdk/pull/3553 a xcm-emulator tests or xcm unit-tests for the coretime-rococo/westend to ensure that `pallet_xcm::reserve_transfer_assets` really works, because you need to probably handle some `Fungible` fees besides `NonFungible(regionId)`, because we need to handle/set `BuyExecution` on the destination, but that is not a problem, I could help, just tell me.

Reply:
> call Broker::create_region_for_benchmark will create region without owner 

Ah, ok, I get it now. Yes, this is a good solution. I was assuming the region would be created with an owner.

Reply:
I merged #3553 into this branch to test if it works, and it did, I managed to run the benchmark and generate weights :)

@bkontur Should we first merge #3553, or have both of these changes in this single PR?

Reply:
@Szegoo very cool, I see we don't even need `create_region_for_benchmark` and Pallet::issue` works, much better :)

> @bkontur Should we first merge https://github.com/paritytech/polkadot-sdk/pull/3553, or have both of these changes in this single PR?

I would go just with one PR